File: src/backend/platform/workflow-policy/contracts/markdownContract.ts
import { existsSync, readFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const REQUIRED_SECTION_NAMES = [
  'TASK_LINEAGE',
  'TASK_METADATA',
  'CONTEXT_PACK_BINDING',
  'REVIEW_OUTCOME',
  'DECISION',
  'DIFFICULTY_LEVEL',
  'RECOMMENDED_EXECUTION',
  'CLOSEOUT_OWNER_AGENT_ID',
  'RETROSPECTIVE_REQUIRED',
] as const;

const REQUIRED_GROUPS = [
  'headingName',
  'labelName',
  'labelValue',
  'title',
  'fenceMarker',
  'fenceLanguage',
] as const;

export type SectionNameKey = typeof REQUIRED_SECTION_NAMES[number];

export interface MarkdownContract {
  version: 1;
  headingRegex: string;
  labelRegex: string;
  titleRegex: string;
  fenceOpenRegex: string;
  groups: Readonly<Record<typeof REQUIRED_GROUPS[number], number>>;
  stripHtmlComments: boolean;
  warnOnDuplicateLabel: boolean;
  opaqueFences: boolean;
  sectionNames: Readonly<Record<SectionNameKey, string>>;
  compiled: Readonly<{
    heading: RegExp;
    label: RegExp;
    title: RegExp;
    fenceOpen: RegExp;
  }>;
}

let cachedContract: MarkdownContract | null = null;
let cachedDefaultPath: string | null = null;

function getDefaultContractPath(): string {
  cachedDefaultPath ??= resolveMarkdownContractPath();
  return cachedDefaultPath;
}

export function resolveMarkdownContractPath(repoRoot = process.cwd()): string {
  let current = path.resolve(repoRoot);
  while (true) {
    const candidate = path.join(current, 'config', 'markdown-contract.default.json');
    if (existsSync(candidate)) {
      return candidate;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      const moduleRelative = path.resolve(
        path.dirname(fileURLToPath(import.meta.url)),
        '../../../../../config/markdown-contract.default.json',
      );
      return existsSync(moduleRelative)
        ? moduleRelative
        : path.join(path.resolve(repoRoot), 'config', 'markdown-contract.default.json');
    }
    current = parent;
  }
}

export function loadMarkdownContract(contractPath = getDefaultContractPath()): MarkdownContract {
  if (cachedContract && contractPath === getDefaultContractPath()) {
    return cachedContract;
  }

  let raw: unknown;
  try {
    raw = JSON.parse(readFileSync(contractPath, 'utf-8'));
  } catch (err) {
    throw new Error(`${contractPath}: invalid JSON: ${err instanceof Error ? err.message : String(err)}`);
  }

  const record = requireRecord(raw, contractPath, '<root>');
  requireField(record.version === 1, contractPath, 'version');
  const headingRegex = requireString(record.headingRegex, contractPath, 'headingRegex');
  const labelRegex = requireString(record.labelRegex, contractPath, 'labelRegex');
  const titleRegex = requireString(record.titleRegex, contractPath, 'titleRegex');
  const fenceOpenRegex = requireString(record.fenceOpenRegex, contractPath, 'fenceOpenRegex');
  const groupsRecord = requireRecord(record.groups, contractPath, 'groups');
  const groups = Object.fromEntries(REQUIRED_GROUPS.map((key) => {
    const value = groupsRecord[key];
    requireField(Number.isInteger(value) && (value as number) > 0, contractPath, `groups.${key}`);
    return [key, value as number];
  })) as Record<typeof REQUIRED_GROUPS[number], number>;

  const sectionNameRecord = requireRecord(record.sectionNames, contractPath, 'sectionNames');
  const sectionNames = Object.fromEntries(REQUIRED_SECTION_NAMES.map((key) => {
    const value = requireString(sectionNameRecord[key], contractPath, `sectionNames.${key}`);
    return [key, value];
  })) as Record<SectionNameKey, string>;

  const compiled = {
    heading: compileRegex(headingRegex, contractPath, 'headingRegex', 'm'),
    label: compileRegex(labelRegex, contractPath, 'labelRegex', 'm'),
    title: compileRegex(titleRegex, contractPath, 'titleRegex', 'm'),
    fenceOpen: compileRegex(fenceOpenRegex, contractPath, 'fenceOpenRegex'),
  };

  smokeTest(contractPath, compiled, groups);

  const contract: MarkdownContract = Object.freeze({
    version: 1,
    headingRegex,
    labelRegex,
    titleRegex,
    fenceOpenRegex,
    groups: Object.freeze(groups),
    stripHtmlComments: record.stripHtmlComments === true,
    warnOnDuplicateLabel: record.warnOnDuplicateLabel === true,
    opaqueFences: record.opaqueFences === true,
    sectionNames: Object.freeze(sectionNames),
    compiled: Object.freeze(compiled),
  });

  if (contractPath === getDefaultContractPath()) {
    cachedContract = contract;
  }
  return contract;
}

export function validateMarkdownContract(contractPath = getDefaultContractPath()): void {
  loadMarkdownContract(contractPath);
}

function requireRecord(value: unknown, contractPath: string, field: string): Record<string, unknown> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error(`${contractPath}: invalid ${field}`);
  }
  return value as Record<string, unknown>;
}

function requireString(value: unknown, contractPath: string, field: string): string {
  if (typeof value !== 'string' || !value) {
    throw new Error(`${contractPath}: invalid ${field}`);
  }
  return value;
}

function requireField(condition: boolean, contractPath: string, field: string): void {
  if (!condition) {
    throw new Error(`${contractPath}: invalid ${field}`);
  }
}

function compileRegex(source: string, contractPath: string, field: string, flags = ''): RegExp {
  try {
    return new RegExp(source, flags);
  } catch (err) {
    throw new Error(`${contractPath}: invalid ${field}: ${err instanceof Error ? err.message : String(err)}`);
  }
}

function smokeTest(
  contractPath: string,
  compiled: MarkdownContract['compiled'],
  groups: MarkdownContract['groups'],
): void {
  if (compiled.heading.exec('##\tTask Lineage ##')?.[groups.headingName] !== 'Task Lineage') {
    throw new Error(`${contractPath}: invalid headingRegex embedded smoke fixture`);
  }
  const labelMatch = compiled.label.exec('- Difficulty Level: Hard <!-- bumped -->');
  if (labelMatch?.[groups.labelName] !== 'Difficulty Level' || labelMatch?.[groups.labelValue] !== 'Hard <!-- bumped -->') {
    throw new Error(`${contractPath}: invalid labelRegex embedded smoke fixture`);
  }
  if (compiled.title.exec('# Task Title ##')?.[groups.title] !== 'Task Title') {
    throw new Error(`${contractPath}: invalid titleRegex embedded smoke fixture`);
  }
  const fenceMatch = compiled.fenceOpen.exec('```bash ');
  if (fenceMatch?.[groups.fenceMarker] !== '```' || fenceMatch?.[groups.fenceLanguage] !== 'bash') {
    throw new Error(`${contractPath}: invalid fenceOpenRegex embedded smoke fixture`);
  }
}

File: src/backend/scripts/python/lib/markdown_contract.py
"""Shared markdown contract loader."""
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path
from types import MappingProxyType
from typing import Mapping, Pattern

REQUIRED_SECTION_NAMES = (
    "TASK_LINEAGE",
    "TASK_METADATA",
    "CONTEXT_PACK_BINDING",
    "REVIEW_OUTCOME",
    "DECISION",
    "DIFFICULTY_LEVEL",
    "RECOMMENDED_EXECUTION",
    "CLOSEOUT_OWNER_AGENT_ID",
    "RETROSPECTIVE_REQUIRED",
)
REQUIRED_GROUPS = (
    "headingName",
    "labelName",
    "labelValue",
    "title",
    "fenceMarker",
    "fenceLanguage",
)


@dataclass(frozen=True)
class CompiledMarkdownContract:
    heading: Pattern[str]
    label: Pattern[str]
    title: Pattern[str]
    fence_open: Pattern[str]


@dataclass(frozen=True)
class MarkdownContract:
    version: int
    heading_regex: str
    label_regex: str
    title_regex: str
    fence_open_regex: str
    groups: Mapping[str, int]
    strip_html_comments: bool
    warn_on_duplicate_label: bool
    opaque_fences: bool
    section_names: Mapping[str, str]
    compiled: CompiledMarkdownContract


def default_markdown_contract_path() -> Path:
    return Path(__file__).resolve().parents[5] / "config" / "markdown-contract.default.json"


@lru_cache(maxsize=8)
def load_markdown_contract(contract_path: str | Path | None = None) -> MarkdownContract:
    path = Path(contract_path) if contract_path is not None else default_markdown_contract_path()
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except Exception as exc:  # noqa: BLE001 - validation reports path and parser message
        raise ValueError(f"{path}: invalid JSON: {exc}") from exc

    if not isinstance(data, dict):
        raise ValueError(f"{path}: invalid <root>")
    _require(data.get("version") == 1, path, "version")
    heading_regex = _string(data.get("headingRegex"), path, "headingRegex")
    label_regex = _string(data.get("labelRegex"), path, "labelRegex")
    title_regex = _string(data.get("titleRegex"), path, "titleRegex")
    fence_open_regex = _string(data.get("fenceOpenRegex"), path, "fenceOpenRegex")
    raw_groups = _mapping(data.get("groups"), path, "groups")
    groups = MappingProxyType({
        key: _positive_int(raw_groups.get(key), path, f"groups.{key}")
        for key in REQUIRED_GROUPS
    })
    raw_names = _mapping(data.get("sectionNames"), path, "sectionNames")
    section_names = MappingProxyType({
        key: _string(raw_names.get(key), path, f"sectionNames.{key}")
        for key in REQUIRED_SECTION_NAMES
    })
    compiled = CompiledMarkdownContract(
        heading=_compile(heading_regex, path, "headingRegex"),
        label=_compile(label_regex, path, "labelRegex"),
        title=_compile(title_regex, path, "titleRegex"),
        fence_open=_compile(fence_open_regex, path, "fenceOpenRegex"),
    )
    _smoke(path, compiled, groups)
    return MarkdownContract(
        version=1,
        heading_regex=heading_regex,
        label_regex=label_regex,
        title_regex=title_regex,
        fence_open_regex=fence_open_regex,
        groups=groups,
        strip_html_comments=data.get("stripHtmlComments") is True,
        warn_on_duplicate_label=data.get("warnOnDuplicateLabel") is True,
        opaque_fences=data.get("opaqueFences") is True,
        section_names=section_names,
        compiled=compiled,
    )


def validate_markdown_contract(contract_path: str | Path | None = None) -> None:
    load_markdown_contract(contract_path)


def _mapping(value: object, path: Path, field: str) -> dict[str, object]:
    if not isinstance(value, dict):
        raise ValueError(f"{path}: invalid {field}")
    return value


def _string(value: object, path: Path, field: str) -> str:
    if not isinstance(value, str) or not value:
        raise ValueError(f"{path}: invalid {field}")
    return value


def _positive_int(value: object, path: Path, field: str) -> int:
    if not isinstance(value, int) or value <= 0:
        raise ValueError(f"{path}: invalid {field}")
    return value


def _require(condition: bool, path: Path, field: str) -> None:
    if not condition:
        raise ValueError(f"{path}: invalid {field}")


def _compile(source: str, path: Path, field: str) -> Pattern[str]:
    try:
        return re.compile(source)
    except re.error as exc:
        raise ValueError(f"{path}: invalid {field}: {exc}") from exc


def _smoke(path: Path, compiled: CompiledMarkdownContract, groups: Mapping[str, int]) -> None:
    if compiled.heading.match("##\tTask Lineage ##").group(groups["headingName"]) != "Task Lineage":
        raise ValueError(f"{path}: invalid headingRegex embedded smoke fixture")
    label = compiled.label.match("- Difficulty Level: Hard <!-- bumped -->")
    if label is None or label.group(groups["labelName"]) != "Difficulty Level" or label.group(groups["labelValue"]) != "Hard <!-- bumped -->":
        raise ValueError(f"{path}: invalid labelRegex embedded smoke fixture")
    if compiled.title.match("# Task Title ##").group(groups["title"]) != "Task Title":
        raise ValueError(f"{path}: invalid titleRegex embedded smoke fixture")
    fence = compiled.fence_open.match("```bash ")
    if fence is None or fence.group(groups["fenceMarker"]) != "```" or fence.group(groups["fenceLanguage"]) != "bash":
        raise ValueError(f"{path}: invalid fenceOpenRegex embedded smoke fixture")
